stan/math/rev/fun/generalized_inverse.hpp

```diff
@@ -0,0 +1,124 @@
+#ifndef STAN_MATH_REV_FUN_GENERALIZED_INVERSE_HPP
+#define STAN_MATH_REV_FUN_GENERALIZED_INVERSE_HPP
+
+#include <stan/math/rev/core.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/rev/fun/cholesky_decompose.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
+#include <stan/math/rev/fun/tcrossprod.hpp>
+#include <stan/math/prim/fun/crossprod.hpp>
+#include <stan/math/prim/fun/inverse_spd.hpp>
+#include <stan/math/rev/fun/mdivide_left_spd.hpp>
+#include <stan/math/prim/fun/mdivide_right_spd.hpp>
+#include <stan/math/rev/fun/inverse.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Reverse mode differentiation algorithm reference:
+ *
+ * The Differentiation of Pseudo-Inverses and Nonlinear Least Squares Problems
+ * Whose Variables Separate. Author(s): G. H. Golub and V. Pereyra. Source: SIAM
+ * Journal on Numerical Analysis, Vol. 10, No. 2 (Apr., 1973), pp. 413-432
+ *
+ * Equation 4.12 in the paper
+ */
+template <typename EigMat, require_rev_matrix_t<EigMat>* = nullptr>
+inline auto generalized_inverse(const EigMat& G) {
+  using value_t = value_type_t<EigMat>;
+
+  if (G.size() == 0)
+    return G;
+
+  const auto n = G.rows();
+  const auto m = G.cols();
+
+  if (G.rows() == G.cols())
+    return inverse(G);
+
+  if (n < m) {
+    arena_t<plain_type_t<EigMat>> G_arena(G);
+    auto A_spd = tcrossprod(G_arena.val());
+    arena_t<EigMat> inv_G(transpose(mdivide_left_spd(A_spd, G_arena.val())));
+    auto aP = (1 - G_arena.val() * inv_G.transpose());
+    auto Pa = (1 - inv_G * G_arena.val());
+    reverse_pass_callback([G_arena, inv_G, aP, Pa]() mutable {
+      G_arena.adj() += -inv_G * G_arena.adj() * inv_G;
+      G_arena.adj() += tcrossprod(inv_G) * G_arena.adj().transpose() * aP;
+      G_arena.adj() += Pa * G_arena.adj().transpose() * crossprod(inv_G());
+    });
+    return inv_G;
```

Comment:
Couple little things here

1. the type returned by `arena_t<>` will always be a `plain_type` so you can do just 

`arena_t<EigMat> G_arena(G);`

2. Anything that goes into the lamda's capture like
```
reverse_pass_callback([%Copied Into Lambda%]() mutable {
```

Needs to be put on our autodiff stack using `arena_t`. For `aP` and `Pa` you can do

```cpp
    auto aP = to_arena((1 - G_arena.val() * inv_G.transpose()));
    auto Pa = to_arena(1 - inv_G * G_arena.val());
```

and then these will be moved over to the arena.

3. Eigen doesn't have the concept of elementwise operations over matrices, so you need to tell eigen to treat things as an array, do the subtraction, then tell eigen to treat it as a matrix like

```cpp
    auto aP = to_arena((1 - (G_arena.val() * inv_G.transpose()).array()).matrix());
    auto Pa = to_arena((1 - (inv_G * G_arena.val()).array()).matrix());
```

4. Since you transpose in 

```
    arena_t<EigMat> inv_G(transpose(mdivide_right_spd(G_arena.val_op(), A_spd)));
```

idt you need to transpose again here

```
    auto aP = to_arena((1 - (G_arena.val_op() * inv_G.val_op().transpose()).array()).matrix());
```

Reply:
> Eigen doesn't have the concept of elementwise operations over matrices

Actually addition and subtraction are supported on matrices. For any other elementwise operation you are correct.

Reply:
I found an implementation in Python and re-read the original paper to find out that it was not correct. I needed to subtract by a diagonal matrix of 1s. I rewrote and now the rev tests are passing!!!

Reply:
Except I get this one error I don't understand,
```
Assertion failed: (dst.rows() == src.rows() && dst.cols() == src.cols()), function resize_if_allowed, file lib/eigen_3.3.9/Eigen/src/Core/AssignEvaluator.h, line 710.
./test/unit/math/mix/fun/generalized_inverse_test --gtest_output="xml:./test/unit/math/mix/fun/generalized_inverse_test.xml" failed
exit now (12/14/20 07:36:31 EST)
```